### PR TITLE
Add agfs fee scheme 12

### PR DESCRIPTION
**Pull request by jsugarman**
#### What
Add new fee scheme 12 for AGFS

#### Ticket

[CBO-1374](https://dsdmoj.atlassian.net/browse/CBO-1374)
[CBO-1405](https://dsdmoj.atlassian.net/browse/CBO-1405)

#### Why
Part of Criminal legal aid review (CLAR)

### CODE REVIEW

File: config/locales/error_messages.en.yml [hunk 1]
```diff
@@ -1035,6 +1035,14 @@ misc_fee:
       long: Enter a valid quantity (1 to 3) for plea and case management hearing fees
       short: Enter a valid quantity (1 to 3)
       api: Enter a valid quantity (1 to 3) for plea and case management hearing fees
+    miumu_numericality:
+      long: Enter a valid quantity (1) for Unused material (upto 3 hours)
+      short: Must be exactly 1
+      api: Enter a valid quantity (3+) for Unused material (over 3 hours)
```
Comment (jsugarman) [inline config/locales/error_messages.en.yml]: wrong message, should be same as `long`

File: db/seeds/schemas/add_agfs_fee_scheme_12.rb [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        @agfs_fee_scheme_12 ||= FeeScheme.agfs.twelve.first
+      end
+
+      def agfs_fee_scheme_11
+        @agfs_fee_scheme_11 ||= FeeScheme.agfs.eleven.first
+      end
+
+      def destroy_agfs_scheme_12_offences
+        if pretending?
+          puts "Would delete #{scheme_12_offence_count} scheme 12 offences".yellow
```
Comment (kmahern) [inline db/seeds/schemas/add_agfs_fee_scheme_12.rb:57]: Debug code? (Can this be removed?)

File: db/seeds/schemas/add_agfs_fee_scheme_12.rb [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+          puts "Would reset offence PK sequence to max id value".yellow
+        else
+          scheme_12_offences = Offence.joins(:fee_schemes).merge(FeeScheme.twelve).merge(FeeScheme.agfs).distinct
+          before_count = scheme_12_offences.count
+          puts "Deleted #{before_count} scheme 12 offences".green if scheme_12_offences.destroy_all
```
Comment (kmahern) [inline db/seeds/schemas/add_agfs_fee_scheme_12.rb:62]: Debug code?

File: lib/rule/struct.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    attr_reader :attribute, :rule_method, :bound, :message
+
+    def initialize(attribute, rule_method, bound, message)
+      @attribute = attribute
+      @rule_method = rule_method
+      @bound = bound
+      @message = message
+    end
+  end
+end
```
Comment (jsugarman) [inline lib/rule/struct.rb:14]: consider just using a struct

File: spec/factories/fee_schemes.rb [hunk 5]
```diff
@@ -19,16 +19,22 @@
 
     trait :agfs_eleven do
       start_date { Date.new(2018, 12, 31).beginning_of_day }
-      end_date { nil }
+      end_date { Date.new(2020, 06, 30).beginning_of_day if Settings.agfs_scheme_12_enabled? }
```
Comment (jsugarman) [inline spec/factories/fee_schemes.rb]: use ` Settings.agfs_scheme_12_release_date.end_of_day - 1.day`

File: spec/lib/rule/set_spec.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      subject(:subsets) { sets.select { |set| set.object.id.eql?(1) } }
+
+      it { is_expected.to all(be_a(described_class)) }
+
+      it 'returns appicable sets' do
+        expect(subsets.flat_map(&:to_a)).to match_array(['rule 1a', 'rule 1b'])
+      end
+    end
+  end
+end
```
Comment (jsugarman) [inline spec/lib/rule/set_spec.rb]: linebreak

File: spec/services/fee_reform/search_offences_spec.rb [hunk 7]
```diff
@@ -13,39 +16,84 @@
...
-      expect(offences.length).to eq(4)
-      expect(offences.map(&:description)).to eq(['Offence 11-4 paTTern', 'Offence 11-3', 'Offence 11-5', 'Offence 11-2'])
+        it 'does not return offences with no matching description, contrary, band or category' do
+          expect(results.map(&:description)).not_to include('Offence 12-1')
+        end
+      end
     end
   end
-end
+end
```
Comment (jsugarman) [inline spec/services/fee_reform/search_offences_spec.rb]: linebreak

File: lib/tasks/scheme_twelve.rake [hunk 8]
```diff
@@ -0,0 +1,27 @@
+require Rails.root.join('db','seeds', 'schemas', 'add_agfs_fee_scheme_12')
+
+namespace :db do
+  namespace :scheme_twelve do
+    desc 'Display status of db structures and records for AGFS fee scheme 12'
+    task :status => :environment do
+      adder = Seeds::Schemas::AddAgfsFeeScheme12.new(pretend: false)
+      puts adder.status
```
Comment (kmahern) [inline lib/tasks/scheme_twelve.rake:8]: Debug code? (can this be removed?)

File: db/seeds/schemas/add_agfs_fee_scheme_12.rb (hunk 2 above)
Comment (jsugarman) [reply db/seeds/schemas/add_agfs_fee_scheme_12.rb:57]: this is output from a rake task

File: db/seeds/schemas/add_agfs_fee_scheme_12.rb (hunk 3 above)
Comment (jsugarman) [reply db/seeds/schemas/add_agfs_fee_scheme_12.rb:62]: [ditto](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/3432#discussion_r463064520)

File: lib/tasks/scheme_twelve.rake (hunk 8 above)
Comment (jsugarman) [reply lib/tasks/scheme_twelve.rake:8]: [ditto](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/3432#discussion_r463064520)

File: spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb [hunk 9]
```diff
@@ -248,6 +247,105 @@ def results(claim)
   end
 end
 
+RSpec.shared_examples 'scheme 12 advocate final claim' do |options|
+  let(:offence) { create(:offence, :with_fee_scheme_eleven) }
```
Comment (kmahern) [inline spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb]: Use fee scheme 12 offence trait?

File: spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb [hunk 10]
```diff
@@ -248,6 +247,105 @@ def results(claim)
...
+      post endpoint(:fees), base_fee_params.merge(claim_id: claim.uuid, fee_type_id: miscellaneous_fee.id), format: :json
+      expect(last_response.status).to eql 201
+
+      post endpoint(:expenses), expense_params.merge(claim_id: claim.uuid, expense_type_id: expense_car.id, distance: 500.38, mileage_rate_id: 1), format: :json
+      expect(last_response.status).to eql 201
+
+      post endpoint(:expenses), expense_params.merge(claim_id: claim.uuid, expense_type_id: expense_hotel.id), format: :json
+      expect(last_response.status).to eql 201
+
+      expect(claim).to be_valid_api_agfs_claim(fee_scheme: ['AGFS', 11], offence: offence, total: 1630.2)
```
Comment (kmahern) [inline spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb]: Should this be expect(claim).to be_valid_api_agfs_claim(fee_scheme: ['AGFS', 12]

File: spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb [hunk 11]
```diff
@@ -248,6 +247,105 @@ def results(claim)
...
+      post endpoint(:fees), base_fee_params.merge(claim_id: claim.uuid, fee_type_id: miscellaneous_fee.id), format: :json
+      expect(last_response.status).to eql 201
+
+      post endpoint(:expenses), expense_params.merge(claim_id: claim.uuid, expense_type_id: expense_car.id, distance: 500.38, mileage_rate_id: 1), format: :json
+      expect(last_response.status).to eql 201
+
+      post endpoint(:expenses), expense_params.merge(claim_id: claim.uuid, expense_type_id: expense_hotel.id), format: :json
+      expect(last_response.status).to eql 201
+
+      expect(claim).to be_valid_api_agfs_claim(fee_scheme: ['AGFS', 11], offence: nil, total: 1630.2)
```
Comment (kmahern) [inline spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb]: Should this be expect(claim).to be_valid_api_agfs_claim(fee_scheme: ['AGFS', 10] if we are specifying Case management system creates a valid scheme 10 fixed fee claim?

File: app/validators/fee/fee_type_rule_validator.rb [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+      @fee_rules = rule_sets.each_with_object([]) do |rule_set, arr|
+        rule_set.each do |rule|
+          arr << Rule::Applicator.new(fee, rule)
+        end
+      end
+    end
+
+    def rule_sets
+      @rule_sets ||= Fee::FeeTypeRules.where(unique_code: fee.fee_type&.unique_code) # TODO: inverse dependency
```
Comment (kmahern) [inline app/validators/fee/fee_type_rule_validator.rb]: Does this TODO still apply? And if so, is it possible to give some more details/clarification about what it relates to?

File: config/locales/error_messages.en.yml [hunk 13]
```diff
@@ -1035,6 +1035,14 @@ misc_fee:
       long: Enter a valid quantity (1 to 3) for plea and case management hearing fees
       short: Enter a valid quantity (1 to 3)
       api: Enter a valid quantity (1 to 3) for plea and case management hearing fees
+    miumu_numericality:
+      long: Enter a valid quantity (1) for Unused material (upto 3 hours)
```
Comment (kmahern) [inline config/locales/error_messages.en.yml]: Not a blocker and super picky but the capitalisation here doesn't seem to match other error messages (e.g. plea and case management hearing fees above). Should the U be lowercase?

File: spec/validators/fee/fee_type_rule_spec.rb [hunk 14]
```diff
@@ -0,0 +1,31 @@
+RSpec.describe Fee::FeeTypeRules, type: :validator do
```
Comment (kmahern) [inline spec/validators/fee/fee_type_rule_spec.rb]: Should this class be fee_type_rules_spec.rb (as the class it tests is fee_type_rules)?

**Review by kmahern — APPROVED** [review]
Some minor comments/suggestions but lgtm 👍

File: spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb (hunk 9 above)
Comment (jsugarman) [reply spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb]: odd, not sure how you saw this but this line is already fixed.

File: spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb (hunk 10 above)
Comment (jsugarman) [reply spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb]: [ditto](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/3432#discussion_r472328726)

File: config/locales/error_messages.en.yml (hunk 13 above)
Comment (jsugarman) [reply config/locales/error_messages.en.yml]: good spot

File: spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb (hunk 11 above)
Comment (jsugarman) [reply spec/api/v1/external_users/claims/integration/advocate_claim_creation_spec.rb]: good catch. this was a leftover from a first iteration and needs to change

File: spec/validators/fee/fee_type_rule_spec.rb (hunk 14 above)
Comment (jsugarman) [reply spec/validators/fee/fee_type_rule_spec.rb]: 👍 absolutely

File: app/validators/fee/fee_type_rule_validator.rb (hunk 12 above)
Comment (jsugarman) [reply app/validators/fee/fee_type_rule_validator.rb]: have inversed the dependency now - this refers to the D in [SOLID principles](https://medium.com/@mari_azevedo/s-o-l-i-d-principles-what-are-they-and-why-projects-should-use-them-50b85e4aa8b6)

**Pull request merged**